DOC: Add hint for "Product space too large"

mobiusklein opened the pull request:
Adds a note to the documentation for `DataFrame.groupby` suggesting that the cause of a `Product space too large to allocate arrays!`  error may be from unexpected behavior when grouping on a combination of sparse categorical columns.

- [ ] closes #xxxx
- [ ] tests added / passed
- [ ] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [ ] whatsnew entry

mobiusklein committed:
Add hint for "Product space too large"

Adds a note to the documentation for `groupby` suggesting that the cause of a `Product space too large to allocate arrays!`  message may be from unexpected behavior when grouping on a combination of sparse categorical columns.

mobiusklein committed:
Remove note from explicit docstring

mobiusklein committed:
Move hint to shared_docs groupby notes section

mobiusklein commented on pandas/core/frame.py:
```diff
@@ -7542,6 +7542,11 @@ def update(
 b   12.3  123.0
 NaN 12.3   33.0
+
+If you encounter ``ValueError("Product space too large to allocate arrays!")``,
```
I've moved this to `pandas.core.shared_docs`'s groupby Notes section.

datapythonista commented:
Feels like it should be better to add this information to the error message itself instead.

mobiusklein committed:
Make note slightly more generic.

pep8speaks commented:
Hello @mobiusklein! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`pandas/core/shared_docs.py`](https://github.com/pandas-dev/pandas/blob/a21aff9d13c8dc94a948d2685aa97503a737afcf/pandas/core/shared_docs.py):

> [Line 148:69](https://github.com/pandas-dev/pandas/blob/a21aff9d13c8dc94a948d2685aa97503a737afcf/pandas/core/shared_docs.py#L148): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace

mobiusklein committed:
Remove trailing whitespace

mobiusklein commented on pandas/core/shared_docs.py:
```diff
@@ -144,6 +144,11 @@
 See the `user guide
 <https://pandas.pydata.org/pandas-docs/stable/groupby.html>`__ for more.
+
+If you encounter ``ValueError("Product space too large to allocate arrays!")``,
```
I've made the text slightly more generic.

rhshadrach commented:
+1 on @datapythonista's comment, I would recommend clarifying the error message itself.

mobiusklein committed:
Update error message with contextual information.

mobiusklein committed:
Correct test expectation.

mobiusklein commented:
I have modified the error message, and adjusted the test to expect the new error message. Note, this error isn't thrown in `groupby` but inside an implementation detail method, which looked generic but apparently is only called by `groupby` right now. If it starts being called by other things, the error message may be misleading.

mobiusklein committed:
Fix PEP8

mobiusklein committed:
PEP8

mobiusklein committed:
PEP8

mroeschke commented on pandas/core/reshape/util.py:
```diff
@@ -40,7 +40,13 @@ def cartesian_product(X):
-        raise ValueError("Product space too large to allocate arrays!")
+        raise ValueError(
+            "Product space too large to allocate arrays! If calling "
+            "`DataFrame.groupby`, you are trying to group over too complex "
```
Instead of changing the error message here, I think it's better if you catch this error wherever it's raising in the groupby code and reraise your new message

```
# in some groupby code

try:
    ....
except ValueError as err:
    # check for this original message
    raise ValueError("You are ...") from err
```

dsaxton commented on pandas/tests/reshape/test_util.py:
```diff
@@ -71,7 +71,13 @@ def test_invalid_input(self, X):
     def test_exceed_product_space(self):
         # GH31355: raise useful error when produce space is too large
-        msg = "Product space too large to allocate arrays!"
+        msg = (
```
I suppose it's okay to match against the full error message here, although the test becomes a tad more brittle

rhshadrach reviewed (changes requested):
This is called from `pandas.core.groupby.groupby.GroupBy._reindex_output`, can you catch the current error there and then throw your more informative one?

Also, can you create a test that uses groupby

mobiusklein commented:
@rhshadrach Thank you for finding the exact call site. I'll do that.

mobiusklein commented:
I will work on this when time permits.

Commit:
Raise more helpful error at call site instead of in generic function

Commit:
style

Commit:
More style

mroeschke commented on pandas/core/shared_docs.py:
```diff
@@ -144,6 +144,12 @@
 See the `user guide
 <https://pandas.pydata.org/pandas-docs/stable/groupby.html>`__ for more.
+
+If you encounter an error regarding memory allocation or space (e.g.
```
Don't think you need this now since this entire message is in the exception.

rhshadrach commented on pandas/core/groupby/groupby.py:
```diff
@@ -2984,9 +2984,17 @@ def _reindex_output(
+            index, _ = MultiIndex.from_product(
+                levels_list, names=self.grouper.names
+            ).sortlevel()
+        except ValueError as err:
+            raise ValueError(
```
Is this the only possible way a ValueError can be raised? We should add a check e.g. `if "Product space too lrage" in str(err):` unless it's really certain.

rhshadrach commented on pandas/core/groupby/groupby.py:
```diff
@@ -2984,9 +2984,17 @@ def _reindex_output(
+        except ValueError as err:
+            raise ValueError(
+                "Product space too large to allocate arrays! You're trying to "
+                "group over too complex a column cross-product to fit in memory. "
```
What does it mean to be complex? I think large may be a better description. Also, if the message were to start with "Group by cross product space", could we remove the entire second sentence?

rhshadrach commented on pandas/core/groupby/groupby.py:
```diff
@@ -2984,9 +2984,17 @@ def _reindex_output(
+            raise ValueError(
+                "Product space too large to allocate arrays! You're trying to "
+                "group over too complex a column cross-product to fit in memory. "
+                "If any of those columns are categorical, you may be able "
```
I think this code is only hit if some column is categorical (L2980-2984 above).

rhshadrach reviewed (changes requested):
Also - need tests for the error message.

Commit:
Make error message more succinct and specific. Added test for error message

rhshadrach commented on pandas/tests/groupby/test_groupby.py:
```diff
@@ -2174,3 +2174,20 @@ def test_groupby_mean_duplicate_index(rand_series_with_duplicate_datetimeindex):
+        v = np.tile(np.arange(25), n // 25)
+        df[chr(i)] = v
+
+    message = r"Group by product space too large to allocate arrays.*"
+    with pytest.raises(ValueError, match=message):
+        g = df.groupby(list(df.columns))
+        g.B.sum()
```
If this doesn't raise a ValueError because of some changes to the code, then I worry what it might to do the CI. Maybe it'd be a good idea to add a timeout context manager to pandas._testing.contexts:

```
@contextmanager
def timeout(seconds):
    def timeout_handler(signum, frame):
        raise RuntimeError

    orig_handler = signal.signal(signal.SIGALRM, timeout_handler)
    signal.alarm(seconds)
    try:
        yield
    finally:
        signal.alarm(0)
        signal.signal(signal.SIGALRM, orig_handler)
```

(Stolen from [SO](https://stackoverflow.com/questions/15018519/python-timeout-context-manager-with-threads))

Then you can do:

```
g = df.groupby(list(df.columns))
with pytest.raises(ValueError, match=message):
    with timeout(1):
        g.B.sum()
```

1 second should be well more than enough, it takes half a millisecond on my machine.

rhshadrach reviewed (changes requested):
Thanks for the update - very much like the new message. Few requests/questions below.

mobiusklein commented on pandas/core/groupby/groupby.py:
```diff
@@ -2988,13 +2988,15 @@ def _reindex_output(
-                "If any of those columns are categorical, you may be able "
-                "to circumvent this issue by passing `observed=True`."
-            ) from err
+        except (ValueError, MemoryError) as err:
```
The memory error is thrown from within NumPy, with the message `Unable to allocate <number> <byte unit> for an array with shape <dims> and data type <dtype>`. By the time we've reached this point, we know we're dealing with a DataFrame with at least one categorical column in the grouped-by columns, and that `observed=False` so memory allocation will be larger with `observed=False` than `observed=True`.

We cannot know if the system is overloaded or not, but we can be certain that if the user passed `observed=True`, it would use less memory. I can add " or increase available memory." to the error message to a cover this case.

rhshadrach replied:
> memory allocation will be larger with observed=False than observed=True

Agreed this is likely to be true but not always. However, in the case it's a reasonable size (whatever that is) and the user is out of memory we don't want to be pointing them down the wrong path. I think we also don't want to be intercepting a MemoryError and replacing it with a ValueError.

Your recommend addition makes sense to me if raise with a MemoryError. I recommend having this PR focus on the "Product space too large" ValueError. Another PR to make the MemoryError more informative would be most welcome, but better to separate.

mobiusklein commented on pandas/core/groupby/groupby.py:
```diff
@@ -2988,13 +2988,15 @@ def _reindex_output(
-            ) from err
+        except (ValueError, MemoryError) as err:
+            err_str = str(err)
+            if "Product space too large" in err_str or "array is too big" in err_str or isinstance(err, MemoryError):
```
"array is too big" is part of the message "array is too big; arr.size * arr.dtype.itemsize is larger than the maximum possible size." which is thrown when the memory allocation overflows `intptr_t`:
https://github.com/numpy/numpy/blob/31e153c4b80b03d0ef0670bd9aa9e52745c7e3fc/numpy/core/src/multiarray/ctors.c#L777-L781. In principle, "a contiguous array of this size is not possible with the memory address space of your computer".

This is distinct from a memory error which is thrown when the memory requested could in theory be allocated but is unavailable (https://github.com/numpy/numpy/blob/31e153c4b80b03d0ef0670bd9aa9e52745c7e3fc/numpy/core/src/multiarray/ctors.c#L828). 

The "product space" error itself is only raised if there is signed integer overflow when computing the dimensions for the cartesian product, implemented at https://github.com/pandas-dev/pandas/blob/059c8bac51e47d6eaaa3e36d6a293a22312925e6/pandas/core/reshape/util.py#L39-L43 in the original issue.

rhshadrach replied:
Thanks - seems to me the only way this can be raised is when the unobserved categoricals run amuck so looks good to me to keep.

Commit:
Split error handling into separate paths.

mobiusklein committed:
Merge branch 'master' into patch-2

Commit:
style

Commit:
style

Commit:
style

rhshadrach reviewed (approved):
lgtm, thanks for the changes!

cc @jreback, especially on the test. I was worried if the code actually tried to allocate the size asked for in the test, so requested a timeout context manager to make sure that if it somehow fails, it at least does so in a graceful manner. ref https://github.com/pandas-dev/pandas/pull/41226#discussion_r659177954

rhshadrach commented:
@mobiusklein - can you merge master. Some failures on the CI, but fairly certain they are unrelated.

Commit:
Merge branch 'master' of https://github.com/pandas-dev/pandas into patch-2

mobiusklein commented:
@rhshadrach I think that there is some sanity checking at a lower level, but the tests may fail on a 32 bit system where even the reduced cross-product space is too large to allocate. I've merged master in and will check where the failures are again.

rhshadrach commented:
@mobiusklein - my apologies! I failed to notice that the context manager I asked to be added does not work on Windows. With this, I think it'd be best to just remove it (but otherwise leave the tests as-is).

Commit:
Cause missing platform resources to be no-op warnings in tests

mobiusklein committed:
Merge branch 'master' into patch-2

Commit:
Pre-commit hooks

Commit:
Missed another alarm call to guard

jreback commented:
if you can merge master will look

Commit:
Merge branch 'master' of https://github.com/pandas-dev/pandas into patch-2

Commit:
Don't suggest setting observable when it is already set

mobiusklein commented:
I've merged in `master` and added a branch case to the test. I was not able to conveniently find the right shape for 32 bit architectures, so settled for at least testing the expectation the call will fail on 32-bit architectures

Commit:
pre-commit

jreback commented:
this PR is way too complicated. I am ok with changing the error message but there is a lot of extraneous stuff here.

mroeschke commented:
Agreed. I think it should be suitable to append to the existing error message to use `observed=True` if data is categorical. Don't think it's completely necessary to test for the extra message

jreback commented:
closing. this would need a much simpler PR to merge.
